<a name="contents"></a>
 [Оглавление](#stage_1)

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание" data-toc-modified-id="Описание-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание</a></span></li><li><span><a href="#Описание-заданий:" data-toc-modified-id="Описание-заданий:-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Описание заданий:</a></span></li><li><span><a href="#Цель-исследования:" data-toc-modified-id="Цель-исследования:-0.3"><span class="toc-item-num">0.3&nbsp;&nbsp;</span>Цель исследования:</a></span></li></ul></li><li><span><a href="#Импортируем-библиотеки" data-toc-modified-id="Импортируем-библиотеки-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Импортируем библиотеки</a></span></li><li><span><a href="#Исследуем-таблицы" data-toc-modified-id="Исследуем-таблицы-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследуем таблицы</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица <code>books</code></a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица <code>authors</code></a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица <code>publishers</code></a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица <code>ratings</code></a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица <code>reviews</code></a></span></li></ul></li><li><span><a href="#Задания:" data-toc-modified-id="Задания:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задания:</a></span><ul class="toc-item"><li><span><a href="#Посчитаем,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитаем,-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитаем, сколько книг вышло <code>после 1 января 2000 года</code></a></span></li><li><span><a href="#Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитаем количество обзоров и среднюю оценку</a></span><ul class="toc-item"><li><span><a href="#ТОП-3-книги-по-количеству-обзоров" data-toc-modified-id="ТОП-3-книги-по-количеству-обзоров-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>ТОП 3 книги по количеству обзоров</a></span></li><li><span><a href="#ТОП-3-книги-по-средней-оценке" data-toc-modified-id="ТОП-3-книги-по-средней-оценке-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>ТОП 3 книги по средней оценке</a></span></li></ul></li><li><span><a href="#Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определим издательство, которое выпустило наибольшее число книг <code>толще 50 страниц</code> — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг — учитывайте только <code>книги с 50 и более оценками</code></a></span></li><li><span><a href="#Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитаем <code>среднее количество обзоров</code> от пользователей, которые поставили <code>больше 50 оценок</code></a></span></li></ul></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></div>

### Описание

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![SQL_13YP.png](https://pictures.s3.yandex.net/resources/scheme_1589269096.png)

### Описание заданий:

- Посчитайте, сколько книг вышло после 1 января 2000 года;  
- Для каждой книги посчитайте количество обзоров и среднюю оценку;  
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;  
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;  
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.  

### Цель исследования:  

- проанализировать базу данных    
- ответить на поставленные вопросы    

## Импортируем библиотеки

<a name="stage_1"></a>

[Наверх к оглавлению](#contents)

In [ ]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine


# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])


# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Исследуем таблицы  

<a name="stage_1"></a>

[Наверх к оглавлению](#contents)

### Таблица `books`

In [ ]:
# посмотрим таблицу books
query = ''' 

SELECT *
FROM books
LIMIT 10 

'''
books = pd.io.sql.read_sql(query, con = engine) 
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

Таблица `books` состоит из данных :  

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `books` связана с таблицами `ratings`, `reviews` и `authors` по полю `book_id`, и с таблицей `publishers` по полю `publisher_id`. 

###  Таблица `authors`

In [ ]:
# таблица authors
query = ''' 

SELECT *
FROM authors
LIMIT 10 

'''
authors = pd.io.sql.read_sql(query, con = engine) 
authors

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

Содержит данные об авторах:  

- `author_id` — идентификатор автора;
- `author` — имя автора.  

Связана с таблицей `books` по `authors_id` 

### Таблица `publishers` 

In [ ]:
# таблица publishers
query = ''' 

SELECT *
FROM publishers
LIMIT 10 

'''
publishers = pd.io.sql.read_sql(query, con = engine) 
publishers

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers` связана с таблицей `books` по `publishers_id`

### Таблица `ratings`

In [ ]:
# таблица ratings
query = ''' 

SELECT *
FROM ratings
LIMIT 10 

'''
ratings = pd.io.sql.read_sql(query, con = engine) 
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;  
- `book_id` — идентификатор книги;  
- `username` — имя пользователя, оставившего оценку;  
- `rating` — оценка книги.  


Таблица `ratings` имеет общие связи с `books` и `reviews` по `book_id`  

### Таблица `reviews`

In [ ]:
# таблица reviews
query = ''' 

SELECT *
FROM reviews
LIMIT 10 

'''
reviews = pd.io.sql.read_sql(query, con = engine) 
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

Содержит данные о пользовательских обзорах на книги:  

- `review_id` — идентификатор обзора;  
- `book_id` — идентификатор книги;  
- `username` — имя пользователя, написавшего обзор;  
- `text` — текст обзора.  

Таблица `reviews` имеет общие связи с `books` и `rating` по `book_id` 

## Задания:

<a name="stage_1"></a>

[Наверх к оглавлению](#contents)

### Посчитаем, сколько книг вышло `после 1 января 2000 года`

In [ ]:
query = ''' 

SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01' 

'''   
count_books = pd.io.sql.read_sql(query, con = engine) 
count_books

count
0    819

После 1 января 2000 года , вышло `819` книг.   

### Для каждой книги посчитаем количество обзоров и среднюю оценку

<a name="stage_1"></a>

[Наверх к оглавлению](#contents)

In [ ]:
query = ''' 

SELECT books.title,
    COUNT(DISTINCT reviews.review_id),
    AVG(ratings.rating)
FROM books
INNER JOIN ratings on ratings.book_id = books.book_id
INNER JOIN reviews on reviews.book_id = books.book_id
GROUP BY books.title 



'''
reviews_and_avg_rating = pd.io.sql.read_sql(query, con = engine)
reviews_and_avg_rating

title  count       avg
0                                         'Salem's Lot      2  3.666667
1                   1 000 Places to See Before You Die      1  2.500000
2    13 Little Blue Envelopes (Little Blue Envelope...      3  4.666667
3    1491: New Revelations of the Americas Before C...      2  4.500000
4                                                 1776      4  4.000000
..                                                 ...    ...       ...
988           Wyrd Sisters (Discworld  #6; Witches #2)      3  3.666667
989                        Xenocide (Ender's Saga  #3)      3  3.400000
990                                    Year of Wonders      4  3.200000
991                        You Suck (A Love Story  #2)      2  4.500000
992  Zen and the Art of Motorcycle Maintenance: An ...      4  3.833333

[993 rows x 3 columns]

Посчитали, `993` книги с разным количеством обзоров и разной средней оценкой    
Но посмотрим `топ 3` по количеству обзоров и средней оценке    

#### ТОП 3 книги по количеству обзоров

In [ ]:
query = ''' 

SELECT books.title,
       COUNT(DISTINCT reviews.review_id),
       AVG(ratings.rating)
FROM books
INNER JOIN ratings on ratings.book_id = books.book_id
INNER JOIN reviews on reviews.book_id = books.book_id
GROUP BY books.title 

ORDER BY COUNT(DISTINCT reviews.review_id) DESC
LIMIT 3;

'''

reviews_and_avg_rating_top_3_count = pd.io.sql.read_sql(query, con = engine)
reviews_and_avg_rating_top_3_count

title  count       avg
0                                Memoirs of a Geisha      8  4.138462
1                            Twilight (Twilight  #1)      7  3.662500
2  Harry Potter and the Chamber of Secrets (Harry...      6  4.287500

#### ТОП 3 книги по средней оценке

In [ ]:
query = ''' 

SELECT books.title,
       COUNT(DISTINCT reviews.review_id),
       AVG(ratings.rating)
FROM books
INNER JOIN ratings on ratings.book_id = books.book_id
INNER JOIN reviews on reviews.book_id = books.book_id
GROUP BY books.title 

ORDER BY AVG(ratings.rating) DESC
LIMIT 3;

'''

reviews_and_avg_rating_top_3_avg = pd.io.sql.read_sql(query, con = engine)
reviews_and_avg_rating_top_3_avg

title  count  avg
0              A Dirty Job (Grim Reaper  #1)      4  5.0
1      A Fistful of Charms (The Hollows  #4)      2  5.0
2  A Woman of Substance (Emma Harte Saga #1)      2  5.0

Как видим, хороший рейтинг книги - еще не показатель для большого количества обзоров и наоборот   
С количеством обзоров, рейтинг может падать, ведь у каждого свой вкус ))   

### Определим издательство, которое выпустило наибольшее число книг `толще 50 страниц` — так вы исключите из анализа брошюры

<a name="stage_1"></a>

[Наверх к оглавлению](#contents)

In [ ]:
query = ''' 

SELECT publishers.publisher,

        COUNT(num_pages) as count_books
        
FROM books
INNER JOIN publishers on publishers.publisher_id = books.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY count_books DESC 
LIMIT 3

'''

thicker_publishers = pd.io.sql.read_sql(query, con = engine)
thicker_publishers

publisher  count_books
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25

Выведено ТОП-3 издательства выпустивших наибольшее число книг, в которых число страниц > 50  
`Penguin Books` на первом месте  

### Определим автора с самой высокой средней оценкой книг — учитывайте только `книги с 50 и более оценками`  

<a name="stage_1"></a>

[Наверх к оглавлению](#contents)

In [ ]:
query = ''' 

SELECT authors.author,
        AVG(ratings.rating) as mean
FROM books
INNER JOIN authors on authors.author_id = books.author_id
INNER JOIN ratings on ratings.book_id = books.book_id
WHERE books.book_id IN (SELECT book_id
                        FROM ratings
                        GROUP BY book_id
                        HAVING COUNT(rating) >= 50)
GROUP BY authors.author
ORDER BY mean DESC
LIMIT 1 


'''    
best_authors = pd.io.sql.read_sql(query, con = engine)
best_authors

author      mean
0  J.K. Rowling/Mary GrandPré  4.287097

Автором с самой высокой средней оценкой (среди книг с 50 и более оценками) , является - `J.K. Rowling/Mary GrandPré`

### Посчитаем `среднее количество обзоров` от пользователей, которые поставили `больше 50 оценок`  

<a name="stage_1"></a>

[Наверх к оглавлению](#contents)

In [ ]:
query = ''' 


SELECT SUM(group_table.count_rev)/COUNT(group_table.count_rev) as mean
        FROM (SELECT reviews.username, 
        COUNT(reviews.username) AS count_rev
FROM reviews
WHERE reviews.username IN (SELECT ratings.username
                            FROM ratings
                            GROUP BY ratings.username
                            HAVING COUNT(ratings.rating_id) > 50)
                  
                            GROUP BY username) AS group_table 
                            

'''         
review_avg = pd.io.sql.read_sql(query, con = engine)
review_avg

mean
0  24.333333

Среднее количество обзоров от пользователей `24.3`  

## Вывод:

<a name="stage_1"></a>

[Наверх к оглавлению](#contents)

* После 1 января 2000 года , вышло 819 книг   
* 993 книги уникальны по просмотрам и рейтингу  
* Хороший рейтинг книги - еще не показатель для большого количества обзоров и наоборот  
* ТОП-3 издательства выпустивших наибольшее число книг, в которых число страниц > 50
`Penguin Books, Vintage Grand, Central Publishing` (рейтинг по убыванию)  
* Автором с самой высокой средней оценкой (среди книг с 50 и более оценками) , является - `J.K. Rowling/Mary GrandPré`  
* Среднее количество обзоров от пользователей `24.3`  

**Рекомендации:**  
* Увеличить базу книг  
* Провести рекламную компанию - приведи друга (по промо приглашению) - по промо будет учитываться кол-во приведенных друзей, на основании чего - возможна определенная скидка. Градация скидки в % это уже отдельная история.  
* Вевести ситему рейтинга пользователей. На самом сайте `допилить` возможность оценки пользователей по прочтенным книгам для публичного обозрения. То есть, у каждого пользователя будет рейтинг в виде условных-виртуальных медалек или совят (еможджи совы - например), чем больше условных медалек, тем выше рейтинг. Что подсознательно будет вызывать у пользователей здоровую конкуренцию к прочтению.

    